File: log_search.rb
# log_search.rb
#!/usr/bin/env ruby
require 'optparse'
require 'stringio'

def highlight(line, pattern, regex)
  if regex
    line.gsub(/#{pattern}/, "\e[33m\\0\e[0m")
  else
    line.gsub(pattern, "\e[33m#{pattern}\e[0m")
  end
end

def search_files(files, pattern, regex, ignore_case, context, line_numbers, count_only, output)
  total_matches = 0
  results = []

  files.each do |filepath|
    unless File.exist?(filepath)
      warn "File not found: #{filepath}"
      next
    end
    lines = File.readlines(filepath, chomp: true)

    if count_only
      match_count = 0
      lines.each do |line|
        if regex
          match_count += 1 if line.match?(Regexp.new(pattern, ignore_case))
        else
          if ignore_case
            match_count += 1 if line.downcase.include?(pattern.downcase)
          else
            match_count += 1 if line.include?(pattern)
          end
        end
      end
      total_matches += match_count
      results << "#{filepath}: #{match_count}"
      next
    end

    matched_indices = []
    lines.each_with_index do |line, idx|
      if regex
        matched_indices << idx if line.match?(Regexp.new(pattern, ignore_case))
      else
        if ignore_case
          matched_indices << idx if line.downcase.include?(pattern.downcase)
        else
          matched_indices << idx if line.include?(pattern)
        end
      end
    end

    if context > 0
      shown = []
      matched_indices.each do |idx|
        (idx - context..idx + context).each do |n|
          shown << n if n >= 0 && n < lines.length
        end
      end
      shown.uniq.sort.each do |idx|
        prefix = matched_indices.include?(idx) ? '>' : ' '
        line = lines[idx]
        line = highlight(line, pattern, regex) if prefix == '>'
        line_num = line_numbers ? "#{idx+1}".rjust(4) : ''
        results << "#{filepath}:#{line_num}: #{line}"
      end
      total_matches += matched_indices.length
    else
      matched_indices.each do |idx|
        line = lines[idx]
        line = highlight(line, pattern, regex)
        line_num = line_numbers ? "#{idx+1}".rjust(4) : ''
        results << "#{filepath}:#{line_num}: #{line}"
      end
      total_matches += matched_indices.length
    end
  end

  if output
    File.open(output, 'w') do |f|
      results.each { |r| f.puts r }
      f.puts "\nTotal matches: #{total_matches}" unless count_only
    end
    puts "Results written to #{output}"
  else
    puts results
    puts "\nTotal matches: #{total_matches}" unless count_only
  end
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: log_search.rb <pattern> <file1> [file2...] [options]"
  opts.on("--regex", "Treat pattern as regex") { options[:regex] = true }
  opts.on("--ignore-case", "Case-insensitive search") { options[:ignore_case] = true }
  opts.on("--context N", Integer, "Show N lines of context") { |v| options[:context] = v }
  opts.on("--line-numbers", "Show line numbers") { options[:line_numbers] = true }
  opts.on("--count", "Show match count only") { options[:count] = true }
  opts.on("--output FILE", "Write results to a file") { |v| options[:output] = v }
end.parse!

files = ARGV
if files.empty? || ARGV.empty?
  puts "Usage: log_search.rb <pattern> <file1> [file2...] [options]"
  exit 1
end

pattern = files.shift
search_files(files, pattern, options[:regex], options[:ignore_case], options[:context] || 0,
             options[:line_numbers], options[:count], options[:output])

File: log_search.go
// log_search.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"regexp"
	"strings"
)

func highlight(line, pattern string, regex bool) string {
	if regex {
		re := regexp.MustCompile(pattern)
		return re.ReplaceAllString(line, "\033[33m$0\033[0m")
	}
	return strings.ReplaceAll(line, pattern, "\033[33m"+pattern+"\033[0m")
}

func searchFiles(files []string, pattern string, regex, ignoreCase bool, context int, lineNumbers, countOnly bool, output string) error {
	var totalMatches int
	var results []string

	flags := 0
	if ignoreCase {
		flags |= regexp.I
	}
	var compiled *regexp.Regexp
	if regex {
		var err error
		compiled, err = regexp.Compile(pattern)
		if err != nil {
			return fmt.Errorf("regex error: %v", err)
		}
	}

	for _, filepath := range files {
		f, err := os.Open(filepath)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Warning: %v\n", err)
			continue
		}
		defer f.Close()
		scanner := bufio.NewScanner(f)
		var lines []string
		for scanner.Scan() {
			lines = append(lines, scanner.Text())
		}
		if err := scanner.Err(); err != nil {
			fmt.Fprintf(os.Stderr, "Error reading %s: %v\n", filepath, err)
		}

		if countOnly {
			matchCount := 0
			for _, line := range lines {
				if regex {
					if compiled.MatchString(line) {
						matchCount++
					}
				} else {
					if ignoreCase {
						if strings.Contains(strings.ToLower(line), strings.ToLower(pattern)) {
							matchCount++
						}
					} else {
						if strings.Contains(line, pattern) {
							matchCount++
						}
					}
				}
			}
			totalMatches += matchCount
			results = append(results, fmt.Sprintf("%s: %d", filepath, matchCount))
			continue
		}

		// Collect matches
		var matchedIndices []int
		for i, line := range lines {
			if regex {
				if compiled.MatchString(line) {
					matchedIndices = append(matchedIndices, i)
				}
			} else {
				if ignoreCase {
					if strings.Contains(strings.ToLower(line), strings.ToLower(pattern)) {
						matchedIndices = append(matchedIndices, i)
					}
				} else {
					if strings.Contains(line, pattern) {
						matchedIndices = append(matchedIndices, i)
					}
				}
			}
		}

		if context > 0 {
			shown := make(map[int]bool)
			for _, idx := range matchedIndices {
				for offset := -context; offset <= context; offset++ {
					n := idx + offset
					if n >= 0 && n < len(lines) {
						shown[n] = true
					}
				}
			}
			for idx := 0; idx < len(lines); idx++ {
				if shown[idx] {
					prefix := " "
					for _, m := range matchedIndices {
						if m == idx {
							prefix = ">"
							break
						}
					}
					line := lines[idx]
					if prefix == ">" {
						line = highlight(line, pattern, regex)
					}
					if lineNumbers {
						results = append(results, fmt.Sprintf("%s:%4d: %s", filepath, idx+1, line))
					} else {
						results = append(results, fmt.Sprintf("%s: %s", filepath, line))
					}
				}
			}
			totalMatches += len(matchedIndices)
		} else {
			for _, idx := range matchedIndices {
				line := lines[idx]
				line = highlight(line, pattern, regex)
				if lineNumbers {
					results = append(results, fmt.Sprintf("%s:%4d: %s", filepath, idx+1, line))
				} else {
					results = append(results, fmt.Sprintf("%s: %s", filepath, line))
				}
			}
			totalMatches += len(matchedIndices)
		}
	}

	if output != "" {
		f, err := os.Create(output)
		if err != nil {
			return err
		}
		defer f.Close()
		for _, r := range results {
			f.WriteString(r + "\n")
		}
		if !countOnly {
			f.WriteString(fmt.Sprintf("\nTotal matches: %d\n", totalMatches))
		}
		fmt.Printf("Results written to %s\n", output)
	} else {
		for _, r := range results {
			fmt.Println(r)
		}
		if !countOnly {
			fmt.Printf("\nTotal matches: %d\n", totalMatches)
		}
	}
	return nil
}

func main() {
	var (
		pattern     = flag.String("pattern", "", "Search pattern")
		regexFlag   = flag.Bool("regex", false, "Treat pattern as regex")
		ignoreCase  = flag.Bool("ignore-case", false, "Case-insensitive search")
		context     = flag.Int("context", 0, "Show N lines of context")
		lineNumbers = flag.Bool("line-numbers", false, "Show line numbers")
		countOnly   = flag.Bool("count", false, "Show match count only")
		output      = flag.String("output", "", "Write results to a file")
	)
	flag.Parse()

	files := flag.Args()
	if *pattern == "" || len(files) == 0 {
		fmt.Println("Usage: log_search <pattern> <file1> [file2...] [options]")
		os.Exit(1)
	}

	if err := searchFiles(files, *pattern, *regexFlag, *ignoreCase, *context, *lineNumbers, *countOnly, *output); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}
